pkg/gofr/datasource/pubsub/nats/client.go

```diff
@@ -0,0 +1,348 @@
+		}
+	}
+}
+
+func (n *NATSClient) fetchAndProcessMessages(ctx context.Context, cons jetstream.Consumer, handler MessageHandler) error {
+	msgs, err := cons.Fetch(n.Config.BatchSize, jetstream.FetchMaxWait(n.Config.MaxWait))
+	if err != nil {
+		return err
+	}
+
+	n.ProcessMessages(ctx, msgs, handler)
+
+	return msgs.Error()
+}
+
+// ProcessMessages processes messages from a consumer.
+func (n *NATSClient) ProcessMessages(ctx context.Context, msgs jetstream.MessageBatch, handler MessageHandler) {
+	for msg := range msgs.Messages() {
+		if err := n.HandleMessage(ctx, msg, handler); err != nil {
+			n.Logger.Errorf("Error handling message: %v", err)
+		}
+	}
+}
+
+// HandleMessage handles a message from a consumer.
+func (n *NATSClient) HandleMessage(ctx context.Context, msg jetstream.Msg, handler MessageHandler) error {
+	if err := handler(ctx, msg); err != nil {
+		n.Logger.Errorf("Error handling message: %v", err)
+		return n.NakMessage(msg)
+	}
+
+	return nil
+}
+
+// NakMessage naks a message from a consumer.
+func (n *NATSClient) NakMessage(msg jetstream.Msg) error {
+	if err := msg.Nak(); err != nil {
+		n.Logger.Errorf("Failed to NAK message: %v", err)
+
+		return err
+	}
+
+	return nil
+}
+
+// HandleFetchError handles fetch errors.
+func (n *NATSClient) HandleFetchError(err error) {
+	n.Logger.Errorf("failed to fetch messages: %v", err)
+	time.Sleep(time.Second) // Backoff on error
+}
+
+// Close closes the NATS client.
+func (n *NATSClient) Close() error {
+	n.subMu.Lock()
+	for _, sub := range n.Subscriptions {
+		sub.Cancel()
+	}
+
+	n.Subscriptions = make(map[string]*Subscription)
+	n.subMu.Unlock()
+
+	if n.Conn != nil {
+		n.Conn.Close()
+	}
+
+	return nil
+}
+
+// DeleteStream deletes a stream in NATS JetStream.
+func (n *NATSClient) DeleteStream(ctx context.Context, name string) error {
+	err := n.Js.DeleteStream(ctx, name)
+	if err != nil {
+		n.Logger.Errorf("failed to delete stream: %v", err)
+
+		return err
+	}
+
+	return nil
+}
+
+// CreateStream creates a stream in NATS JetStream.
+func (n *NATSClient) CreateStream(ctx context.Context, cfg StreamConfig) error {
+	n.Logger.Debugf("Creating stream %s", cfg.Stream)
+	jsCfg := jetstream.StreamConfig{
+		Name:     cfg.Stream,
+		Subjects: cfg.Subjects,
+	}
+
+	_, err := n.Js.CreateStream(ctx, jsCfg)
+	if err != nil {
+		n.Logger.Errorf("failed to create stream: %v", err)
+
+		return err
+	}
+
+	return nil
+}
+
+// CreateOrUpdateStream creates or updates a stream in NATS JetStream.
+func (n *NATSClient) CreateOrUpdateStream(ctx context.Context, cfg *jetstream.StreamConfig) (jetstream.Stream, error) {
+	n.Logger.Debugf("Creating or updating stream %s", cfg.Name)
+
+	stream, err := n.Js.CreateOrUpdateStream(ctx, *cfg)
+	if err != nil {
+		n.Logger.Errorf("failed to create or update stream: %v", err)
+
+		return nil, err
+	}
+
+	return stream, nil
+}
+
+// ValidateConfigs validates the configuration for NATS JetStream.
+func ValidateConfigs(conf *Config) error {
```

Comment:
need it exported to test

Reply:
Which test are we talking about here?

Reply:
As soon as you change the package name of your test, for example, `package client`, to `package client_test`, you now are limited to only having access to exported functions from your test.

Reply:
I don't know what you're asking for here, do you want to eliminate this test and reduce test coverage? If you change `client_test.go` from `nats_test` package to `nats` that is not go idiomatic and would break linters.